fendermint/app/src/app.rs

```diff
@@ -291,6 +299,37 @@ where
         let state = self.committed_state()?;
         Ok((state.state_params, state.block_height))
     }
+
+    fn verify_ipc_message(&self, ipc_message: IPCMessage) -> bool {
+        match ipc_message {
+            IPCMessage::TopDown(finality) => self.parent_finality.check_finality(&finality),
+            IPCMessage::BottomUp => unimplemented!(),
+        }
+    }
+
+    /// Verifies the list of transactions to see if they are valid. Returns true if all the txns
+    /// are valid and false if any one of the txns is invalid.
+    fn verify_messages(&self, txns: &Vec<Bytes>) -> bool {
+        for tx in txns {
+            let is_ok = match serde_json::from_slice::<ChainMessage>(tx) {
+                Err(_) => {
+                    // FIXME: maybe we can reject the proposal when the transaction cannot be deserialized
+                    tracing::info!("cannot deserialized transaction: {tx:?}");
```

Comment:
This was also my first intuition, but then I thought we can also let the proposal pass and punish the validator during execution.